### PR TITLE
POC for table filters with codecs for findOne

### PR DESCRIPTION
**Pull request by amorton**
NOTE: this PR builds on the #1309 `ajm/validate-poc` PR, it should apply to main after that so change the target to main once that PR is merged. If not I can rebase. 

CI https://github.com/stargate/data-api/actions/runs/10102611929

**What this PR does**:

This PR refactors how we process filters by moving the base class FilterableResolver that was used by command resolvers to be standalone class FilterResolver that has CollectionFilterResolver and TableFilterResolver subclasses. The changes to the resolvers are mostly to handle this change, other than findOne.

The next part is findOne updated to process the filter for a table. Using the TableFilterResolver to make filters, which then use the NativeTypeTableFilter and JSONCodec to map from the types used for JSOn into what CQL expects.

This is POC work, we then need to expand this to handle all the situations we expect.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [*] Changes manually tested
- [ ] Automated Tests added/updated
- [ ] Documentation added/updated
- [*] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/builder/BuiltConditionPredicate.java [hunk 1]
```diff
@@ -10,15 +10,16 @@ public enum BuiltConditionPredicate {
   IN("IN"),
   CONTAINS("CONTAINS"),
   NOT_CONTAINS("NOT CONTAINS"),
-  CONTAINS_KEY("CONTAINS KEY"),
-  ;
+  CONTAINS_KEY("CONTAINS KEY");
 
-  private final String cql;
+  public final String cql;
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/builder/BuiltConditionPredicate.java:15]: Why is this made public?

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/builder/BuiltConditionPredicate.java [hunk 2]
```diff
@@ -10,15 +10,16 @@ public enum BuiltConditionPredicate {
...
 
   BuiltConditionPredicate(String cql) {
     this.cql = cql;
   }
 
+  // TIDY - remove this use of toString() it should be used for log msg's etc, not core
+  // functionality. This is called to build the CQL string we execute.
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/builder/BuiltConditionPredicate.java:22]: Agreed, `toString()` is for diagnostics etc, not for "real" use cases.

I have no strong opinions b/w opening up public String field (as done here) or separate access method.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/TableFilter.java [hunk 3]
```diff
@@ -1,11 +1,42 @@
...
+import java.util.List;
 
+/**
+ * A {@link DBFilterBase} that is applied to a table (i.e. not a Collection) to filter the rows to
+ * read.
+ */
 public abstract class TableFilter extends DBFilterBase {
-  // TODO- the path is the column name here, maybe rename ?
+
+  // TIDY - the path is the column name here, maybe rename ?
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/TableFilter.java:15]: Yes, change to `column`.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/codecs/JSONCodec.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    /**
+     * Returns an instance that just returns the value passed in, the same as {@link
+     * Function#identity()}
+     *
+     * <p>Unsafe because it does not catch any errors from the conversion, because there are none.
+     *
+     * @return
+     * @param <T>
+     */
+    static <T> FromJava<T, T> unsafeIdentity() {
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/codecs/JSONCodec.java:111]: And... what is the point here? Is it for type-casting purpose or why is this needed?

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/codecs/JSONCodecRegistry.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+ *
+ * <p>IMPORTANT: There must be a codec for every CQL data type we want to write to, even if the
+ * translation is an identity translation. This is so we know if the translation can happen, and
+ * then if it was done correctly with the actual value. See {@link
+ * JSONCodec.FromJava#unsafeIdentity()} for the identity mapping, and example usage in {@link #TEXT}
+ * codec.
+ */
+public class JSONCodecRegistry {
+
+  // Internal list of all codes
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/codecs/JSONCodecRegistry.java:25]: typo: codes -> codecs

**Review by tatu-at-datastax — COMMENTED** [review]
Looks ok to me, but would want @maheshrajamani to also review.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/TableFilterResolver.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  private static final Object DYNAMIC_TEXT_GROUP = new Object();
+  private static final Object DYNAMIC_NUMBER_GROUP = new Object();
+
+  public TableFilterResolver(OperationsConfig operationsConfig) {
+    super(operationsConfig);
+  }
+
+  @Override
+  protected FilterMatchRules<T> buildMatchRules() {
+    var matchRules = new FilterMatchRules<T>();
```
Comment (maheshrajamani) [inline src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/TableFilterResolver.java:31]: We should make the match rules an Application Scoped as this doesn't change by find class for table.

Comment (maheshrajamani) [reply src/main/java/io/stargate/sgv2/jsonapi/service/resolver/matcher/TableFilterResolver.java:31]: Also need to do the same in `CollectionFilterResolver`

**Review by maheshrajamani — APPROVED** [review]

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/NativeTypeTableFilter.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    throw new UnsupportedOperationException(
+        "No supported - will be modified when we migrate collections filters java driver");
+  }
+
+  @Override
+  public Select apply(
+      TableSchemaObject tableSchemaObject, Select select, List<Object> positionalValues) {
+
+    // TODO: AARON return the correct errors, this is POC work now
+    var column =
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/NativeTypeTableFilter.java:101]: +1 if doable.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/FindTableOperation.java [hunk 8]
```diff
@@ -35,13 +48,32 @@ public FindTableOperation(
...
+
+    // Start the select
+    Select select =
+        selectFrom(
+                commandContext.schemaObject().tableMetadata.getKeyspace(),
+                commandContext.schemaObject().tableMetadata.getName())
+            .json()
+            .all(); // TODO: this is where we would do the field selection / projection
+
+    // BUG: this probably break order for nested expressions, for now enough to get this tested
```
Comment (Yuqi-Du) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/FindTableOperation.java:60]: Leave the BUG comment here, 
Seems like here we are not doing any real filters here, POC work.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/FindTableOperation.java [hunk 9]
```diff
@@ -35,13 +48,32 @@ public FindTableOperation(
...
+
+    // Add the where clause operations
+    List<Object> positonalValues = new ArrayList<>();
+    for (TableFilter tableFilter : tableFilters) {
+      select = tableFilter.apply(commandContext.schemaObject(), select, positonalValues);
+    }
+
+    select = select.limit(params.limit());
+
+    // Building a statment using the positional values added by the TableFilter
```
Comment (Yuqi-Du) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/FindTableOperation.java:75]: statement

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/codecs/JSONCodecRegistry.java (hunk 5 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/codecs/JSONCodecRegistry.java:25]: solved

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/FindTableOperation.java (hunk 9 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/FindTableOperation.java:75]: solved

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/TableFilter.java (hunk 3 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/TableFilter.java:15]: solved

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/NativeTypeTableFilter.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+  public BuiltCondition get() {
+    throw new UnsupportedOperationException(
+        "No supported - will be modified when we migrate collections filters java driver");
+  }
+
+  @Override
+  public Select apply(
+      TableSchemaObject tableSchemaObject, Select select, List<Object> positionalValues) {
+
+    // TODO: AARON return the correct errors, this is POC work now
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/NativeTypeTableFilter.java:100]: solved

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/codecs/JSONCodec.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+   * JSONCodecRegistry}
+   *
+   * @param <T> The type of the Java object that needs to be transformed into the type CQL expects
+   * @param <R> The type Java object the CQL driver expects
+   */
+  @FunctionalInterface
+  public interface FromJava<T, R> {
+
+    /**
+     * Convers the current Java value to the type CQL expects.
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/codecs/JSONCodec.java:92]: solved

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/codecs/JSONCodec.java (hunk 4 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/codecs/JSONCodec.java:111]: added a TODO comment

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/codecs/JSONCodec.java [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+     *
+     * @param t
+     * @param targetType The type of the CQL column the value will be written to, passed so it can
+     *     be used when creating an exception if there was a error doing the transformation.
+     * @return
+     * @throws FromJavaCodecException
+     */
+    R apply(T t, DataType targetType) throws FromJavaCodecException;
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/codecs/JSONCodec.java:100]: agree, changed.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/NativeTypeTableFilter.java [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        tableSchemaObject
+            .tableMetadata
+            .getColumn(path)
+            .orElseThrow(() -> new IllegalArgumentException("Column not found: " + path));
+
+    var codec =
+        JSONCodecRegistry.codecFor(column.getType(), columnValue)
+            .orElseThrow(
+                () ->
+                    new RuntimeException(
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/NativeTypeTableFilter.java:111]: solved

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/NativeTypeTableFilter.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/NativeTypeTableFilter.java:101]: add a todo comment here.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/NativeTypeTableFilter.java [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(NativeTypeTableFilter.class);
+
+  /**
+   * The operations that can be performed to filter a column TIDY: we have operations defined in
+   * multiple places, once we have refactored the collection operations we should centralize these
+   * operator definitions
+   */
+  public enum Operator {
+    EQ(BuiltConditionPredicate.EQ),
```
[the remark this answers is not in the document]
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/NativeTypeTableFilter.java:57]: added a TODO comment

**Pull request merged**